src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -210,6 +215,34 @@ public KeyRangeIterator.Builder getIndexQueryResults(Collection<Expression> expr
         return builder;
     }
 
+    private void maybeTriggerGuardrails(QueryViewBuilder.QueryView queryView)
+    {
+        if (!cfs.isLeveledCompaction())
+            return;
+
+        boolean nonPartitionRestricted = false;
+
+        for (Pair<Expression, Collection<SSTableIndex>> pair : queryView.view)
+        {
+            if (pair.left.getIndexTermType().columnMetadata().kind != ColumnMetadata.Kind.PARTITION_KEY)
+            {
+                nonPartitionRestricted = true;
+                break;
+            }
+        }
+
+        if (!nonPartitionRestricted)
+            return;
+
+        Guardrails.maximumSaiSSTablesWithNonParitionRestrictedQuery.guard(queryView.referencedIndexes.size(),
+                                                                          format("%s on columns %s", command.metadata().toString(),
+                                                                                 queryView.view.stream()
+                                                                                               .map(p -> p.left.getIndexTermType().columnMetadata().name.toString())
+                                                                                               .collect(Collectors.joining(", "))),
+                                                                          false,
+                                                                          null); // ???? how to propagate ClientState here?
```

Comment:
this is pretty tricky!

Reply:
See `SortedTableWriter#guardParititonThreshold()`. It doesn't actually use the client state, but passes null there.